ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,525 @@
+-- chain contains at most @k@ blocks in the Praos Race Window anchored at the
+-- the intersection.
+--
+-- Definition of the /Races in Stability Windows Assumption/.
+-- Every adversarial chain has at most @k@ blocks in every Praos Race Window of
+-- the honest chain ending within 'Delta' slots of the Stability Window
+-- starting at the slot of the first adversarial block. (Crucially, these
+-- are the Praos Race Windows in which the adversary could not yet have
+-- drastically accelerated its rate of election.)
+--
+-- Definition of the /Genesis Implication Conjecture/. We conjecture that
+-- assuming that the Genesis window length is no greater than the Stability
+-- Window and that every possible adversarial chain satisfies the Races
+-- in Stability Windows Assumption, then the honest chain strictly wins every possible density
+-- comparison from the Ouroboros Genesis paper. The key intuition is that a less
+-- dense chain would have to lose at least one race within the Genesis window.
+checkAdversarialChain ::
+  forall base hon adv.
+     AdversarialRecipe base hon
+  -> ChainSchema base adv
+  -> Exn.Except (AdversarialViolation hon adv) ()
+checkAdversarialChain recipe adv = do
+    checkStart
+    checkCount
+    checkRaces
+  where
+    AdversarialRecipe {
+        arHonest = ChainSchema winH vH
+      ,
+        arParams = (Kcp k, Scg s, Delta d)
+      ,
+        arPrefix
+      } = recipe
+
+    ChainSchema winA vA = adv
+
+    checkStart = do
+        let startA       = C.windowStart winA :: C.Index base SlotE
+            intersection = startA C.- 1       :: C.Index base SlotE
+
+        case C.toWin winH intersection of
+            Nothing -> do
+              -- genesis block is the only permissible anchor outside of @hon@
+              when (startA   /= C.Count 0) $ Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+              when (arPrefix /= C.Count 0) $ Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+            Just i -> do
+                let _ = i :: C.Index hon SlotE
+
+                when (BV.testV S.inverted vH i) $ do
+                    Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+
+                C.SomeWindow Proxy precedingSlots <-
+                    pure $ C.withWindowBetween (C.windowSize winH) (C.Lbl @"foo") (C.Count 0) i
+                let pc = BV.countActivesInV S.notInverted (C.sliceV precedingSlots vH)
+
+                when (C.frWinVar precedingSlots (C.toVar pc) /= arPrefix) $ do
+                    Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+    checkCount = do
+        let pc = BV.countActivesInV S.notInverted vA
+        when (C.toVar pc <= 0) $ Exn.throwError BadCount
+
+    -- the youngest slot in the adversarial schedule that cannot have accelerated
+    --
+    -- (IE s past the first active adversarial slot.)
+    youngestStableA :: C.Index adv SlotE
```

Comment:
The first slot after the stability window can have a leader resulting from an accelerated leader schedule IIUC. So something is off with the name of this variable, or its value needs to be corrected.